src/abstractarray.jl

```diff
@@ -120,103 +104,67 @@ similar{SA<:SizedArray,T,S}(::Type{SA},::Type{T},s::Size{S}) = sizedarray_simila
 similar{A<:Array,T,S}(::Type{A},::Type{T},s::Size{S}) = sizedarray_similar_type(T,s,length_val(s))()
 
 
-# This is used in Base.LinAlg quite a lot, and it impacts type stability
-# since some functions like expm() branch on a check for Hermitian or Symmetric
-# TODO much more work on type stability. Base functions are using similar() with
-# size, which poses difficulties!!
-@generated function Base.full{T,SM<:StaticMatrix}(sym::Symmetric{T,SM})
-    exprs_up = [i <= j ? :(m[$(sub2ind(size(SM), i, j))]) : :(m[$(sub2ind(size(SM), j, i))]) for i = 1:size(SM,1), j=1:size(SM,2)]
-    exprs_lo = [i >= j ? :(m[$(sub2ind(size(SM), i, j))]) : :(m[$(sub2ind(size(SM), j, i))]) for i = 1:size(SM,1), j=1:size(SM,2)]
-
-    return quote
-        $(Expr(:meta, :inline))
-        m = sym.data
-        if sym.uplo == 'U'
-            @inbounds return SM($(Expr(:call, SM, Expr(:tuple, exprs_up...))))
-        else
-            @inbounds return SM($(Expr(:call, SM, Expr(:tuple, exprs_lo...))))
-        end
+@inline reshape(a::StaticArray, s::Size) = similar_type(a, s)(Tuple(a))
+@generated function reshape(a::AbstractArray, s::Size{S}) where {S}
+    if IndexStyle(a) == IndexLinear() # TODO this isn't "hyperpure"
```

Comment:
Can you make this safer by dispatching on the `IndexStyle` in "regular" functions?

```julia
reshape(a::AbstractArray, s::Size) = _reshape(IndexStyle(a), a, s)
@generated function _reshape(::IndexLinear, a, s::Size{S}) where {S}
...
end
@generated function _reshape(::IndexStyle, a, s::Size{S}) where {S}
...
end
```